File: dart/05-functions/functions_higher_order.dart
// A higher-order function: it takes another function as a parameter.
List<int> applyToAll(List<int> items, int Function(int) transform) {
  return items.map(transform).toList();
}

// A function that returns a function (a function factory)
int Function(int) multiplier(int factor) {
  return (int x) => x * factor;
}

void main() {
  var numbers = [1, 2, 3, 4, 5];

  // Pass a named function as an argument
  int double(int n) => n * 2;
  print('Doubled: ${applyToAll(numbers, double)}');

  // Pass an anonymous function (lambda) directly
  print('Squared: ${applyToAll(numbers, (n) => n * n)}');

  // Build a specialized function and use it
  var triple = multiplier(3);
  print('Tripled: ${applyToAll(numbers, triple)}');

  // Built-in higher-order methods
  var evens = numbers.where((n) => n.isEven).toList();
  var sum = numbers.reduce((a, b) => a + b);
  print('Evens: $evens');
  print('Sum: $sum');
}

File: javascript/05-functions/functions_higher_order.js
const numbers = [1, 2, 3, 4, 5, 6];

// map: transform every element
const doubled = numbers.map((n) => n * 2);

// filter: keep elements that pass a test
const evens = numbers.filter((n) => n % 2 === 0);

// reduce: collapse a list into a single value
const total = numbers.reduce((acc, n) => acc + n, 0);

// A function that returns a function (a "factory")
function multiplier(factor) {
  return (n) => n * factor;
}

const triple = multiplier(3);

console.log("doubled:", doubled);
console.log("evens:", evens);
console.log("total:", total);
console.log("triple(7) =", triple(7));
